lib/oxigraph/src/sparql/dataset.rs

```diff
@@ -1,18 +1,235 @@
+        // not present. Assuming a URI.
+        _ => {
+            // Note that Term::from_str() will not work for URIs
+            // (OxRDF NamedNode) when the string is not within "<"
+            // and ">" delimiters.
+            let named_node = match NamedNodeRef::new(*Arc::from(s)) {
+                Ok(n) => n,
+                Err(e) => {
+                    return Err(Error::new(
+                        ErrorKind::InvalidData,
+                        format!("iri parse error {e} for {s}"),
+                    ))
+                }
+            };
+            match Term::from_str(&named_node.to_string()) {
+                Ok(s) => Ok(s),
+                Err(e) => Err(Error::new(
+                    ErrorKind::InvalidData,
+                    format!("iri parse error {e} for {s}"),
+                )),
+            }
+        }
+    }
+}
+
+/// Convert triple string formats from OxRDF to HDT.
+fn term_to_hdt_bgp_str(term: &Term) -> Result<String, StorageError> {
+    let hdt_str = match term {
+        // Remove double quote delimiters from URIs.
+        Term::NamedNode(named_node) => named_node.clone().into_string(),
+
+        // Get the string directly from literals and add
+        // quotes to work-around handling of "\n" being
+        // double-escaped.
+        // format!("\"{}\"", literal.value()),
+        Term::Literal(literal) => {
+            if literal.is_plain() {
+                literal.to_string().replace("\\n", "\n")
+            }
+            // For numbers and other typed literals return
+            // None as the BGP search will need to collect
+            // all possibilities before filtering.
+            else {
+                return Err(StorageError::Corruption(CorruptionError::new(format!(
+                    "unhandled literal value {literal}"
+                ))));
+            }
+        }
+
+        Term::BlankNode(_s) => term.to_string(),
+
+        // Otherwise use the string directly.
+        Term::Triple(_) => term.to_string(),
+    };
+
+    Ok(hdt_str)
+}
+
+impl QueryableDataset for HDTDatasetView {
+    type InternalTerm = String;
+    type Error = StorageError;
+
+    fn internal_quads_for_pattern(
+        &self,
+        subject: Option<&String>,
+        predicate: Option<&String>,
+        object: Option<&String>,
+        graph_name: Option<Option<&String>>,
+    ) -> Box<dyn Iterator<Item = Result<InternalQuad<Self>, StorageError>>> {
+        if let Some(graph_name) = graph_name {
+            if graph_name.is_some() {
+                return Box::new(
+                    vec![Err(StorageError::Corruption(CorruptionError::new(
+                        format!("HDT does not support named graph: {graph_name:?}"),
+                    )))]
+                    .into_iter(),
+                );
+            }
+        }
+        let s = subject.cloned();
+        let p = predicate.cloned();
+        let o = object.cloned();
```

Comment:
Genuine question: do you need these clones?

Reply:
You can do `s.map(|s| s.as_str())` to convert from `Option<&String>` to `Option<&str>` without reallocating.

Reply:
Or shorter yet, `s.as_deref()`, c.f. [`Option::as_deref`](https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.as_deref).